File: pyvulkan/22 - Indexed Rendering/start/descriptors.py
from config import *
import vklogging

class DescriptorSetLayoutData:


    def __init__(self):

        self.count = 0
        self.indices = []
        self.types = []
        self.counts = []
        self.stages = []

def make_descriptor_set_layout(device, bindings: DescriptorSetLayoutData):

    layoutBindings = []

    for i in range(bindings.count):

        """
        typedef struct VkDescriptorSetLayoutBinding {
            uint32_t              binding;
            VkDescriptorType      descriptorType;
            uint32_t              descriptorCount;
            VkShaderStageFlags    stageFlags;
            const VkSampler*      pImmutableSamplers;
        } VkDescriptorSetLayoutBinding;
        """
        
        layoutBinding = VkDescriptorSetLayoutBinding(
            binding = bindings.indices[i],
            descriptorType = bindings.types[i],
            descriptorCount = bindings.counts[i],
            stageFlags = bindings.stages[i]
        )

        layoutBindings.append(layoutBinding)
    
    """
    typedef struct VkDescriptorSetLayoutCreateInfo {
        VkStructureType                        sType;
        const void*                            pNext;
        VkDescriptorSetLayoutCreateFlags       flags;
        uint32_t                               bindingCount;
        const VkDescriptorSetLayoutBinding*    pBindings;
    } VkDescriptorSetLayoutCreateInfo;
    """

    layoutInfo = VkDescriptorSetLayoutCreateInfo(
        bindingCount = bindings.count,
        pBindings = layoutBindings
    )

    try:
        return vkCreateDescriptorSetLayout(device, layoutInfo, None)
    except:
        vklogging.logger.print("Failed to create Descriptor Set Layout")
        return None

def make_descriptor_pool(device, size: int, bindings: DescriptorSetLayoutData):

    poolSizes = []
    """
    typedef struct VkDescriptorPoolSize {
        VkDescriptorType    type;
        uint32_t            descriptorCount;
    } VkDescriptorPoolSize;
    """

    for _type in bindings.types:

        poolSize = VkDescriptorPoolSize(
            type = _type,
            descriptorCount = size
        )

        poolSizes.append(poolSize)
    
    """
        typedef struct VkDescriptorPoolCreateInfo {
            VkStructureType                sType;
            const void*                    pNext;
            VkDescriptorPoolCreateFlags    flags;
            uint32_t                       maxSets;
            uint32_t                       poolSizeCount;
            const VkDescriptorPoolSize*    pPoolSizes;
        } VkDescriptorPoolCreateInfo;
    """

    poolInfo = VkDescriptorPoolCreateInfo(
        poolSizeCount = len(poolSizes),
        pPoolSizes = poolSizes,
        maxSets = size
    )

    try:
        return vkCreateDescriptorPool(device, poolInfo, None)
    except:
        vklogging.logger.print("Failed to make descriptor pool")
        return None

def allocate_descriptor_set(device, descriptorPool, descriptorSetLayout):

    """
        typedef struct VkDescriptorSetAllocateInfo {
			VkStructureType                 sType;
			const void*                     pNext;
			VkDescriptorPool                descriptorPool;
			uint32_t                        descriptorSetCount;
			const VkDescriptorSetLayout*    pSetLayouts;
		} VkDescriptorSetAllocateInfo;
    """

    allocationInfo = VkDescriptorSetAllocateInfo(
        descriptorPool = descriptorPool, descriptorSetCount = 1, 
        pSetLayouts = [descriptorSetLayout,]
    )

    try:
        return vkAllocateDescriptorSets(device, allocationInfo)[0]
    except:
        vklogging.logger.print("Failed to allocate a descriptor set")
        return None

File: vulkan 2022/fast raycaster/view/vkInit/descriptors.cpp
#include "descriptors.h"
#include "../../control/logging.h"


vk::DescriptorSetLayout vkInit::make_descriptor_set_layout(
	vk::Device device, const descriptorSetLayoutData& bindings) {

	/*
		Bindings describes a whole bunch of descriptor types, collect them all into a
		list of some kind.
	*/
	std::vector<vk::DescriptorSetLayoutBinding> layoutBindings;
	layoutBindings.reserve(bindings.count);

	for (int i = 0; i < bindings.count; i++) {

		/*
			typedef struct VkDescriptorSetLayoutBinding {
				uint32_t              binding;
				VkDescriptorType      descriptorType;
				uint32_t              descriptorCount;
				VkShaderStageFlags    stageFlags;
				const VkSampler*      pImmutableSamplers;
			} VkDescriptorSetLayoutBinding;
		*/

		vk::DescriptorSetLayoutBinding layoutBinding;
		layoutBinding.binding = bindings.indices[i];
		layoutBinding.descriptorType = bindings.types[i];
		layoutBinding.descriptorCount = bindings.counts[i];
		layoutBinding.stageFlags = bindings.stages[i];
		layoutBindings.push_back(layoutBinding);
	}

	/*
		typedef struct VkDescriptorSetLayoutCreateInfo {
			VkStructureType                        sType;
			const void*                            pNext;
			VkDescriptorSetLayoutCreateFlags       flags;
			uint32_t                               bindingCount;
			const VkDescriptorSetLayoutBinding*    pBindings;
		} VkDescriptorSetLayoutCreateInfo;
	*/
	vk::DescriptorSetLayoutCreateInfo layoutInfo;
	layoutInfo.flags = vk::DescriptorSetLayoutCreateFlagBits();
	layoutInfo.bindingCount = bindings.count;
	layoutInfo.pBindings = layoutBindings.data();


	try {
		return device.createDescriptorSetLayout(layoutInfo);
	}
	catch (vk::SystemError err) {

		vkLogging::Logger::get_logger()->print("Failed to create Descriptor Set Layout");

		return nullptr;
	}
}

vk::DescriptorPool vkInit::make_descriptor_pool(
	vk::Device device, uint32_t size, const descriptorSetLayoutData& bindings) {

	std::vector<vk::DescriptorPoolSize> poolSizes;
	/*
		typedef struct VkDescriptorPoolSize {
			VkDescriptorType    type;
			uint32_t            descriptorCount;
		} VkDescriptorPoolSize;
	*/

	for (int i = 0; i < bindings.count; i++) {

		vk::DescriptorPoolSize poolSize;
		poolSize.type = bindings.types[i];
		poolSize.descriptorCount = size;
		poolSizes.push_back(poolSize);
	}

	vk::DescriptorPoolCreateInfo poolInfo;
	/*
		typedef struct VkDescriptorPoolCreateInfo {
			VkStructureType                sType;
			const void*                    pNext;
			VkDescriptorPoolCreateFlags    flags;
			uint32_t                       maxSets;
			uint32_t                       poolSizeCount;
			const VkDescriptorPoolSize*    pPoolSizes;
		} VkDescriptorPoolCreateInfo;
	*/

	poolInfo.flags = vk::DescriptorPoolCreateFlags();
	poolInfo.maxSets = size;
	poolInfo.poolSizeCount = static_cast<uint32_t>(poolSizes.size());
	poolInfo.pPoolSizes = poolSizes.data();

	try {
		return device.createDescriptorPool(poolInfo);
	}
	catch (vk::SystemError err) {
		vkLogging::Logger::get_logger()->print("Failed to make descriptor pool");
		return nullptr;
	}
}

vk::DescriptorSet vkInit::allocate_descriptor_set(
	vk::Device device, vk::DescriptorPool descriptorPool,
	vk::DescriptorSetLayout layout) {

	vk::DescriptorSetAllocateInfo allocationInfo;
	/*
		typedef struct VkDescriptorSetAllocateInfo {
			VkStructureType                 sType;
			const void*                     pNext;
			VkDescriptorPool                descriptorPool;
			uint32_t                        descriptorSetCount;
			const VkDescriptorSetLayout*    pSetLayouts;
		} VkDescriptorSetAllocateInfo;
	*/

	allocationInfo.descriptorPool = descriptorPool;
	allocationInfo.descriptorSetCount = 1;
	allocationInfo.pSetLayouts = &layout;

	try {
		return device.allocateDescriptorSets(allocationInfo)[0];
	}
	catch (vk::SystemError err) {
		vkLogging::Logger::get_logger()->print("Failed to allocate descriptor set from pool");
		return nullptr;
	}
}